doc/xorg-docs/general/fonts/fonts.7
.\" automatically generated with docbook2mdoc fonts.xml
.Dd 16 March 2012
.Dt FONTS 7
.Os
.Sh NAME
.Nm fonts
.Nd Fonts in X11R6
.Sh INTRODUCTION
This document describes the support for fonts in X11R6.
.Sx Installing_fonts
is aimed at the
casual user wishing to install fonts in X11R6 the rest of the
document describes the font support in more detail.
.Pp
We assume some familiarity with digital fonts.
If anything is not
clear to you, please consult
.Sx Appendix_background_and_terminology
at the
end of this document for background information.
.Ss Two font systems
X11 includes two font systems: the original core X11 fonts
system, which is present in all implementations of X11, and the Xft
fonts system, which may not yet be distributed with implementations of
X11 that are not based on either XFree86 or X11R6.8 or later.
.Pp
The core X11 fonts system is directly derived from the fonts system
included with X11R1 in 1987, which could only use monochrome bitmap
fonts.
Over the years, it has been more or less happily coerced into
dealing with scalable fonts and rotated glyphs.
.Pp
Xft was designed from the start to provide good support for scalable
fonts, and to do so efficiently.
Unlike the core fonts system, it
supports features such as anti-aliasing and sub-pixel rasterisation.
Perhaps more importantly, it gives applications full control over the
way glyphs are rendered, making fine typesetting and WYSIWIG display
possible.
Finally, it allows applications to use fonts that are not
installed system-wide for displaying documents with embedded fonts.
.Pp
Xft is not compatible with the core fonts system: usage of Xft
requires fairly extensive changes to toolkits (user-interface
libraries).  While X.Org will continue to maintain the core fonts
system, toolkit authors are encouraged to switch to Xft as soon as
possible.
.Sh INSTALLING FONTS
This section explains how to configure both Xft and the core fonts
system to access newly-installed fonts.
.Ss Configuring Xft
Xft has no configuration mechanism itself, it relies upon the
.Lk http://www.fontconfig.org/ fontconfig
library to configure and customise fonts.
That library is
not specific to the X Window system, and does not rely on any
particular font output mechanism.
.Pp
.Sy Installing fonts in Xft
.Pp
Fontconfig looks for fonts in a set of well-known directories that
include all of X11R6's standard font directories
.Pf ( Pa /usr/share/fonts/X11/* )
by default) as well as a
directory called
.Pa .fonts/
in the user's home directory.
Installing a font for use by Xft applications is as simple
as copying a font file into one of these directories.
.Bd -literal
$ cp lucbr.ttf ~/.fonts/
.Ed
.Pp
Fontconfig will notice the new font at the next opportunity and rebuild its
list of fonts.
If you want to trigger this update from the command
line, you may run the command
.Dq Nm fc-cache .
.Bd -literal
$ fc-cache
.Ed
.Pp
In order to globally update the system-wide Fontconfig information on
Unix systems, you will typically need to run this command as root:
.Bd -literal
$ su -c fc-cache
.Ed
.Pp
.Sy Fine-tuning Xft
.Pp
Fontconfig's behaviour is controlled by a set of configuration
files: a standard configuration file,
.Pa /etc/fonts/fonts.conf ,
a host-specific configuration file,
.Pa /etc/fonts/local.conf ,
and a user-specific file called
.Pa .fonts.conf
in the user's
home directory (this can be overridden with the
.Dq Ev FONTCONFIG_FILE
environment variable).
.Pp
Every Fontconfig configuration file must start with the following
boilerplate:
.Bd -literal
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
.Ed
.Pp
In addition, every Fontconfig configuration file must end with the
following line:
.Bd -literal
</fontconfig>
.Ed
.Pp
The default Fontconfig configuration file includes the directory
.Pa \[u02DC]/.fonts/
in the list of directories searched for font
files, and this is where user-specific font files should be installed.
In the unlikely case that a new font directory needs to be added, this
can be done with the following syntax:
.Bd -literal
<dir>/usr/local/share/fonts/</dir>
.Ed
.Pp
Another useful option is the ability to disable anti-aliasing (font
smoothing) for selected fonts.
This can be done with the following
syntax:
.Bd -literal
<match target="font">
<test qual="any" name="family">
<string>Lucida Console</string>
</test>
<edit name="antialias" mode="assign">
<bool>false</bool>
</edit>
</match>
.Ed
.Pp
Anti-aliasing can be disabled for all fonts by the following incantation:
.Bd -literal
<match target="font">
<edit name="antialias" mode="assign">
<bool>false</bool>
</edit>
</match>
.Ed
.Pp
Xft supports sub-pixel rasterisation on LCD displays.
X11R6 should
automatically enable this feature on laptops and when using an LCD
monitor connected with a DVI cable; you can check whether this was
done by typing
.Bd -literal
$ xdpyinfo -ext RENDER | grep sub-pixel
.Ed
.Pp
If this doesn't print anything, you will need to configure Render for
your particular LCD hardware manually; this is done with the following
syntax:
.Bd -literal
<match target="font">
<edit name="rgba" mode="assign">
<const>rgb</const>
</edit>
</match>
.Ed
.Pp
The string
.Dq Li rgb
within the
.Dq Li <const> . . . Ns Dq Ns Li </const>
specifies the order of pixel components on your display, and should be
changed to match your hardware; it can be one of
.Dq Li rgb
(normal
LCD screen),
.Dq Li bgr
(backwards LCD screen),
.Dq Li vrgb
(LCD
screen rotated clockwise) or
.Dq Li vbgr
(LCD screen rotated
counterclockwise).
.Pp
.Sy Configuring applications
.Pp
A growing number of applications use Xft in preference to the core
fonts system.
Some applications, however, need to be explicitly
configured to use Xft.
.Pp
A case in point is XTerm, which can be set to use Xft by using the
.Dq Li -fa
command line option or by setting the
.Dq Li XTerm*faceName
resource:
.Bd -literal
XTerm*faceName: Courier
.Ed
.Pp
or
.Bd -literal
$ xterm -fa "Courier"
.Ed
.Pp
For KDE applications, you should select
.Dq Anti-alias fonts
in the
.Dq Fonts
panel of KDE's
.Dq Control Center .
Note that this option is
misnamed: it switches KDE to using Xft but doesn't enable
anti-aliasing in case it was disabled by your Xft configuration file.
.Pp
Gnome applications and Mozilla Firefox will use Xft by default.
.Ss Configuring the core X11 fonts system
Installing fonts in the core system is a two step process.
First,
you need to create a
.Em font directory
that contains all the
relevant font files as well as some index files.
You then need to
inform the X server of the existence of this new directory by
including it in the
.Em font path .
.Pp
.Sy Installing bitmap fonts
.Pp
The X11R6 server can use bitmap fonts in both the cross-platform
BDF format and the somewhat more efficient binary PCF format.
(X11R6 also supports the obsolete SNF format.)
.Pp
Bitmap fonts are normally distributed in the BDF format.
Before
installing such fonts, it is desirable (but not absolutely necessary)
to convert the font files to the PCF format.
This is done by using the
command
.Dq Nm bdftopcf ,
.Em e.g.
.Bd -literal
$ bdftopcf courier12.bdf
.Ed
.Pp
You may then want to compress the resulting PCF font files:
.Bd -literal
$ gzip courier12.pcf
.Ed
.Pp
After the fonts have been converted, you should copy all the font
files that you wish to make available into a arbitrary directory, say
.Pa /usr/local/share/fonts/bitmap/ .
You should then create the
index file
.Pa fonts.dir
by running the command
.Dq Nm mkfontdir
(please see the
.Lk mkfontdir.1.html mkfontdir(1)
manual page for more information):
.Bd -literal
$ mkdir /usr/local/share/fonts/bitmap/
$ cp *.pcf.gz /usr/local/share/fonts/bitmap/
$ mkfontdir /usr/local/share/fonts/bitmap/
.Ed
.Pp
All that remains is to tell the X server about the existence of the
new font directory; see
.Sx Setting_the_servers_font_path
below.
.Pp
.Sy Installing scalable fonts
.Pp
The X11R6 server supports scalable fonts in multiple
formats, including Type\ 1, TrueType, and OpenType/CFF.
(Earlier versions of X11 also included support for the Speedo and
CID scalable font formats, but that is not included in current releases.)
.Pp
Installing scalable fonts is very similar to installing bitmap fonts:
you create a directory with the font files, and run
.Dq Nm mkfontdir
to create an index file called
.Pa fonts.dir .
.Pp
There is, however, a big difference:
.Dq Nm mkfontdir
cannot
automatically recognise scalable font files.
For that reason, you
must first index all the font files in a file called
.Pa fonts.scale .
While this can be done by hand, it is best done
by using the
.Dq Nm mkfontscale
utility.
.Bd -literal
$ mkfontscale /usr/local/share/fonts/Type1/
$ mkfontdir /usr/local/share/fonts/Type1/
.Ed
.Pp
Under some circumstances, it may be necessary to modify the
.Pa fonts.scale
file generated by
.Nm mkfontscale ;
for more
information, please see the
.Lk mkfontdir.1.html mkfontdir(1)
and
.Lk mkfontscale.1.html mkfontscale(1)
manual pages and
.Sx Core_fonts_and_internationalisation
later in this document.
.Pp
.Sy CID-keyed fonts
.Pp
The CID-keyed font format was designed by Adobe Systems for fonts
with large character sets.
The CID-keyed format is obsolete, as it
has been superseded by other formats such as OpenType/CFF and
support for CID-keyed fonts has been removed from X11.
.Pp
.Sy Setting the server's font path
.Pp
The list of directories where the server looks for fonts is known
as the
.Em font path .
Informing the server of the existence of a new
font directory consists of putting it on the font path.
.Pp
The font path is an ordered list; if a client's request matches
multiple fonts, the first one in the font path is the one that gets
used.
When matching fonts, the server makes two passes over the font
path: during the first pass, it searches for an exact match; during
the second, it searches for fonts suitable for scaling.
.Pp
For best results, scalable fonts should appear in the font path before
the bitmap fonts; this way, the server will prefer bitmap fonts to
scalable fonts when an exact match is possible, but will avoid scaling
bitmap fonts when a scalable font can be used.
(The
.Dq Li :unscaled
hack, while still supported, should no longer be necessary in X11R6.)
.Pp
You may check the font path of the running server by typing the command
.Bd -literal
$ xset q
.Ed
.Pp
.Sy Font path catalogue directories
.Pp
You can specify a special kind of font path directory in the form
.Pa catalogue:<dir> .
The directory specified after the
.Pa catalogue:
prefix will be scanned for symlinks  and  each  symlink destination will be
added as a local font path entry.
.Pp
The symlink can be suffixed by attributes such as
.Pf ' Ql unscaled Ns ',
which will be passed through
to the underlying font path entry.
The only exception is the newly
introduced
.Pf ' Ql pri Ns '
attribute, which will be
used for ordering the font paths specified by the symlinks.
.Pp
An example configuration:
.Bd -literal
         75dpi:unscaled:pri=20 ->  /usr/share/X11/fonts/75dpi
         ghostscript:pri=60 ->  /usr/share/fonts/default/ghostscript
         misc:unscaled:pri=10 ->  /usr/share/X11/fonts/misc
         type1:pri=40 ->  /usr/share/X11/fonts/Type1
         type1:pri=50 ->  /usr/share/fonts/default/Type1
.Ed
.Pp
This will add
.Pa /usr/share/X11/fonts/misc
as the
first font path entry with the attribute
.Ql unscaled .
This is functionally equivalent to
setting the following font path:
.Bd -literal
         /usr/share/X11/fonts/misc:unscaled,
         /usr/share/X11/fonts/75dpi:unscaled,
         /usr/share/X11/fonts/Type1,
         /usr/share/fonts/default/Type1,
         /usr/share/fonts/default/ghostscript
.Ed
.Pp
.Sy Temporary modification of the font path
.Pp
The
.Dq Nm xset
utility may be used to modify the font path for the
current session.
The font path is set with the command
.Nm xset fp ;
a new element is added to the front with
.Nm xset +fp ,
and added to
the end with
.Nm xset fp+ .
For example,
.Bd -literal
$ xset +fp /usr/local/fonts/Type1
$ xset fp+ /usr/local/fonts/bitmap
.Ed
.Pp
Conversely, an element may be removed from the front of the font path
with
.Dq Nm xset -fp ,
and removed from the end with
.Dq Nm xset fp- .
You may reset the font path to its default value with
.Dq Nm xset fp default .
.Pp
For more information, please consult the
.Lk xset.1.html xset(1)
manual page.
.Pp
.Sy Permanent modification of the font path
.Pp
The default font path (the one used just after server startup or
after
.Dq Nm xset fp default )
may be specified in the
X server's
.Pa xorg.conf
file.
It is computed by appending all the
directories mentioned in the
.Dq Li FontPath
entries of the
.Dq Li Files
section in the order in which they appear.
If no font path is specified in a config file, the server uses a default
value specified when it was built.
.Bd -literal
FontPath "/usr/local/fonts/Type1"
\&...
FontPath "/usr/local/fonts/bitmap"
.Ed
.Pp
For more information, please consult the
.Lk xorg.conf.5.html xorg.conf(5)
manual page.
.Pp
.Sy Troubleshooting
.Pp
If you seem to be unable to use some of the fonts you have
installed, the first thing to check is that the
.Pa fonts.dir
files
are correct and that they are readable by the server (the X server
usually runs as root, beware of NFS-mounted font directories).  If
this doesn't help, it is quite possible that you are trying to use a
font in a format that is not supported by your server.
.Pp
X11R6 supports the BDF, PCF, SNF, Type 1, TrueType, and OpenType
font formats.
However, not all X11R6 servers
come with all the font backends configured in.
.Pp
On most platforms, the X11R6 servers no longer uses font
backends from modules that are loaded at runtime.
The built in
font support corresponds to the functionality formerly provided by
these modules:
.Bl -bullet
.It
.Ql \(dqbitmap\(dq :
bitmap fonts
.Pf ( Pa *.bdf ,
.Pa *.pcf
and
.Pa *.snf ) ;
.It
.Ql \(dqfreetype\(dq :
TrueType fonts
.Pf ( Pa *.ttf
and
.Pa *.ttc ) ,
OpenType fonts
.Pf ( Pa *.otf
and
.Pa *.otc )
and
Type\ 1 fonts
.Pf ( Pa *.pfa
and
.Pa *.pfb ) .
.El
.Sh FONTS INCLUDED WITH X11R6
.Ss Standard bitmap fonts
The Sample Implementation of X11 (SI) comes with a large number of
bitmap fonts, including the
.Dq Li fixed
family, and bitmap versions
of Courier, Times, Helvetica and some members of the Lucida family.
.Pp
In X11R6, a number of these fonts are provided in Unicode-encoded
font files now.
At build time, these fonts are split into font
files encoded according to legacy encodings, a process which allows
us to provide the standard fonts in a number of regional encodings
with no duplication of work.
.Pp
For example, the font file
.Bd -literal
/usr/share/fonts/X11/misc/6x13.bdf
.Ed
.Pp
with XLFD
.Bd -literal
-misc-fixed-medium-r-semicondensed--13-120-75-75-c-60-iso10646-1
.Ed
.Pp
is a Unicode-encoded version of the standard
.Dq Li fixed
font with
added support for the Latin, Greek, Cyrillic, Georgian, Armenian, IPA
and other scripts plus numerous technical symbols.
It contains over
2800 glyphs, covering all characters of ISO\ 8859 parts 1-5,
7-10, 13-15, as well as all European IBM and Microsoft code pages,
KOI8, WGL4, and the repertoires of many other character sets.
.Pp
This font is used at build time for generating the font files
.Bd -literal
6x13-ISO8859-1.bdf
6x13-ISO8859-2.bdf
\&...
6x13-ISO8859-15.bdf
6x13-KOI8-R.bdf
.Ed
.Pp
with respective XLFDs
.Bd -literal
-misc-fixed-medium-r-normal--13-120-75-75-c-60-iso8859-1
\&...
-misc-fixed-medium-r-normal--13-120-75-75-c-60-iso8859-15
-misc-fixed-medium-r-normal--13-120-75-75-c-60-koi8-r
.Ed
.Pp
The standard short name
.Dq Li fixed
is normally an alias for
.Bd -literal
-misc-fixed-medium-r-normal--13-120-75-75-c-60-iso8859-1
.Ed
.Ss The ClearlyU Unicode font family
The ClearlyU family of fonts provides a set of 12\ pt,
100\ dpi proportional fonts with many of the glyphs needed for
Unicode text.
Together, the fonts contain approximately 7500 glyphs.
.Pp
The main ClearlyU font has the XLFD
.Bd -literal
-mutt-clearlyu-medium-r-normal--17-120-100-100-p-101-iso10646-1
.Ed
.Pp
and resides in the font file
.Bd -literal
/usr/share/fonts/X11/misc/cu12.pcf.gz
.Ed
.Pp
Additional ClearlyU fonts include
.Bd -literal
-mutt-clearlyu alternate glyphs-medium-r-normal--17-120-100-100-p-91-iso10646-1
-mutt-clearlyu pua-medium-r-normal--17-120-100-100-p-111-iso10646-1
-mutt-clearlyu arabic extra-medium-r-normal--17-120-100-100-p-103-fontspecific-0
-mutt-clearlyu ligature-medium-r-normal--17-120-100-100-p-141-fontspecific-0
.Ed
.Pp
The
.Em Alternate Glyphs
font contains additional glyph shapes that
are needed for certain languages.
A second alternate glyph font will
be provided later for cases where a character has more than one
commonly used alternate shape
.Pf ( Em e.g.
the Urdu heh).
.Pp
The
.Em PUA
font contains extra glyphs that are useful for certain
rendering purposes.
.Pp
The
.Em Arabic Extra
font contains the glyphs necessary for
characters that don't have all of their possible shapes encoded in
ISO\ 10646.
The glyphs are roughly ordered according to the order
of the characters in the ISO\ 10646 standard.
.Pp
The
.Em Ligature
font contains ligatures for various scripts that
may be useful for improved presentation of text.
.Ss Standard scalable fonts
X11R6 includes all the scalable fonts distributed with X11R6.
.Pp
.Sy Standard Type\e1 fonts
.Pp
The IBM Courier set of fonts cover ISO\ 8859-1 and
ISO\ 8859-2 as well as Adobe Standard Encoding.
These fonts have
XLFD
.Bd -literal
-adobe-courier-medium-*-*--0-0-0-0-m-0-*-*
.Ed
.Pp
and reside in the font files
.Bd -literal
/usr/share/fonts/X11/Type1/cour*.pfa
.Ed
.Pp
The Adobe Utopia set of fonts only cover ISO\ 8859-1 as well as
Adobe Standard Encoding.
These fonts have XLFD
.Bd -literal
-adobe-utopia-*-*-normal--0-0-0-0-p-0-iso8859-1
.Ed
.Pp
and reside in the font files
.Bd -literal
/usr/share/fonts/X11/Type1/UT*.pfa
.Ed
.Pp
Finally, X11R6 also comes with Type\ 1 versions of Bitstream
Courier and Charter.
These fonts have XLFD
.Bd -literal
-bitstream-courier-*-*-normal--0-0-0-0-m-0-iso8859-1
-bitstream-charter-*-*-normal--0-0-0-0-p-0-iso8859-1
.Ed
.Pp
and reside in the font files
.Bd -literal
/usr/share/fonts/X11/Type1/c*bt_.pfb
.Ed
.Ss The Bigelow & Holmes Luxi family
X11R6 includes the
.Em Luxi
family of scalable fonts, in both
TrueType and Type\ 1 format.
This family consists of the fonts
.Em Luxi Serif ,
with XLFD
.Bd -literal
-b&h-luxi serif-medium-*-normal--*-*-*-*-p-*-*-*
.Ed
.Pp
.Em Luxi Sans ,
with XLFD
.Bd -literal
-b&h-luxi sans-medium-*-normal--*-*-*-*-p-*-*-*
.Ed
.Pp
and
.Em Luxi Mono ,
with XLFD
.Bd -literal
-b&h-luxi mono-medium-*-normal--*-*-*-*-m-*-*-*
.Ed
.Pp
Each of these fonts comes Roman, oblique, bold and bold oblique variants
The TrueType version have glyphs covering the basic ASCII Unicode
range, the Latin\ 1 range, as well as the
.Em Extended Latin
range and some additional punctuation characters.
In particular,
these fonts include all the glyphs needed for ISO\ 8859 parts 1,
2, 3, 4, 9, 13 and 15, as well as all the glyphs in the Adobe Standard
encoding and the Windows 3.1 character set.
.Pp
The glyph coverage of the Type\ 1 versions is somewhat reduced,
and only covers ISO\ 8859 parts 1, 2 and 15 as well as the Adobe
Standard encoding.
.Pp
The Luxi fonts are original designs by Kris Holmes and Charles
Bigelow.
Luxi fonts include seriffed, sans serif, and monospaced
styles, in roman and oblique, and normal and bold weights.
The fonts
share stem weight, x-height, capital height, ascent and descent, for
graphical harmony.
.Pp
The character width metrics of Luxi roman and bold fonts match those
of core fonts bundled with popular operating and window systems.
.Pp
The license terms for the Luxi fonts are included in the file
.Pa COPYRIGHT.BH ,
as well as in the
.Lk License "License document"
.Pq Bigelow_Holmes_Inc_and_URW_GmbH_Luxi_font_license .
.Pp
Charles Bigelow and Kris Holmes from Bigelow and Holmes Inc.
developed the Luxi typeface designs in Ikarus digital format.
.Pp
URW++ Design and Development GmbH converted the Ikarus format fonts
to TrueType and Type1 font programs and implemented the grid-fitting
"hints" and kerning tables in the Luxi fonts.
.Pp
For more information, please contact
.Aq Mt design@bigelowandholmes.com
or
.Aq Mt info@urwpp.de ,
or consult
.Lk http://www.urwpp.de "the URW++ web site" .
.Pp
An earlier version of the Luxi fonts was made available under the
name
.Em Lucidux .
This name should no longer be used due to
trademark uncertainties, and all traces of the
.Em Lucidux
name have been removed from X11R6.
.Sh MORE ABOUT CORE FONTS
This section describes XFree86-created enhancements to the core
X11 fonts system that were adopted by X.Org.
.Ss Core fonts and internationalisation
The scalable font backends (Type\ 1 and TrueType) can
automatically re-encode fonts to the encoding specified in the
XLFD in
.Pa fonts.dir .
For example, a
.Pa fonts.dir
file can
contain entries for the Type\ 1 Courier font such as
.Bd -literal
cour.pfa -adobe-courier-medium-r-normal--0-0-0-0-m-0-iso8859-1
cour.pfa -adobe-courier-medium-r-normal--0-0-0-0-m-0-iso8859-2
.Ed
.Pp
which will lead to the font being recoded to ISO\ 8859-1 and
ISO\ 8859-2 respectively.
.Pp
.Sy The fontenc layer
.Pp
Two of the scalable backends (Type\ 1 and the
.Em FreeType
TrueType backend) use a common
.Em fontenc
layer for
font re-encoding.
This allows these backends to share their encoding
data, and allows simple configuration of new locales independently of
font type.
.Pp
.Em Please note:
the X-TrueType (X-TT) backend is not included
in X11R6.  That functionality has been merged into the FreeType
backend.
.Pp
In the
.Em fontenc
layer, an encoding is defined by a name (such as
.Ql iso8859-1 ) ,
possibly a number of aliases (alternate names), and
an ordered collection of mappings.
A mapping defines the way the
encoding can be mapped into one of the
.Em target encodings
known to
.Em fontenc ;
currently, these consist of Unicode, Adobe glyph names,
and arbitrary TrueType
.Dq cmap Ns s.
.Pp
A number of encodings are hardwired into
.Em fontenc ,
and are
therefore always available; the hardcoded encodings cannot easily be
redefined.
These include:
.Bl -bullet
.It
.Ql iso10646-1 :
Unicode;
.It
.Ql iso8859-1 :
ISO\ Latin-1 (Western Europe);
.It
.Ql iso8859-2 :
ISO\ Latin-2 (Eastern Europe);
.It
.Ql iso8859-3 :
ISO\ Latin-3 (Southern Europe);
.It
.Ql iso8859-4 :
ISO\ Latin-4 (Northern Europe);
.It
.Ql iso8859-5 :
ISO\ Cyrillic;
.It
.Ql iso8859-6 :
ISO\ Arabic;
.It
.Ql iso8859-7 :
ISO\ Greek;
.It
.Ql iso8859-8 :
ISO\ Hebrew;
.It
.Ql iso8859-9 :
ISO\ Latin-5 (Turkish);
.It
.Ql iso8859-10 :
ISO\ Latin-6 (Nordic);
.It
.Ql iso8859-15 :
ISO\ Latin-9, or Latin-0 (Revised
Western-European);
.It
.Ql koi8-r :
KOI8 Russian;
.It
.Ql koi8-u :
KOI8 Ukrainian (see RFC 2319);
.It
.Ql koi8-ru :
KOI8 Russian/Ukrainian;
.It
.Ql koi8-uni :
KOI8
.Dq Unified

xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 83a98543b58c661a22574a6f8d8f9d777c0955b8)

commit 9eaa0b5c89f645c50ea90e07e8b49ebacf4349a0
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:00:05 2011 -0500

    xace: ricer tuning for XaceHook
    
    gcc doesn't want to hoist the check for XaceHooks[hook] != NULL above the
    varargs code for some reason, so do it ourselves.
    
    Before:
    40000000 trep @   0.0010 msec (1050420.2/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (1921147.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 11977fab546da35d579ebe285e3c26864007805e)

commit a3b6ecc893e17028cfbe16d8d7981c9037fc0766
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Tue Jan 3 10:45:45 2012 +1000

    xfree86: split warning about missing identifier or input driver
    
    Check for identifier first and bail if it's missing (also remove the current

app/xterm/vttests/fonts.sh
#!/bin/sh
# $XTermId: fonts.sh,v 1.11 2011/12/11 16:21:22 tom Exp $
# -----------------------------------------------------------------------------
# this file is part of xterm
#
# Copyright 1999-2003,2011 by Thomas E. Dickey
# 
#                         All Rights Reserved
# 
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE ABOVE LISTED COPYRIGHT HOLDER(S) BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
# 
# Except as contained in this notice, the name(s) of the above copyright
# holders shall not be used in advertising or otherwise to promote the
# sale, use or other dealings in this Software without prior written
# authorization.
# -----------------------------------------------------------------------------
# Demonstrate the use of dynamic colors by setting the background successively
# to different values.

ESC=""
CMD='/bin/echo'
OPT='-n'
SUF=''
TMP=/tmp/xterm$$
eval '$CMD $OPT >$TMP || echo fail >$TMP' 2>/dev/null
( test ! -f $TMP || test -s $TMP ) &&
for verb in printf print ; do
    rm -f $TMP
    eval '$verb "\c" >$TMP || echo fail >$TMP' 2>/dev/null
    if test -f $TMP ; then
	if test ! -s $TMP ; then
	    CMD="$verb"
	    OPT=
	    SUF='\c'
	    break
	fi
    fi
done
rm -f $TMP

exec </dev/tty
old=`stty -g`
stty raw -echo min 0  time 5

$CMD $OPT "${ESC}]50;?${SUF}" > /dev/tty
read original

stty $old
original="${original}${SUF}"

if ( trap "echo exit" EXIT 2>/dev/null ) >/dev/null
then
    trap '$CMD $OPT "$original" >/dev/tty; exit' EXIT HUP INT TRAP TERM
else
    trap '$CMD $OPT "$original" >/dev/tty; exit' 0    1   2   5    15
fi

F=1
D=1
T=6
while true
do
    $CMD $OPT "${ESC}]50;#$F${SUF}" >/dev/tty
    #sleep 1
    if test .$D = .1 ; then
	test $F = $T && D=-1
    else
	test $F = 1 && D=1
    fi
    F=`expr $F + $D`
done
